src/V3Ast.h

```diff
@@ -1814,8 +1814,18 @@ class AstNode VL_NOT_FINAL {
     // METHODS - Tree modifications
     // Returns nodep, adds newp to end of nodep's list
     static AstNode* addNext(AstNode* nodep, AstNode* newp);
+    template <typename T_NodeLeft, typename T_NodeRight>
+    static auto addNext(T_NodeLeft* nodep, T_NodeRight* newp) {
+        return reinterpret_cast<typename std::common_type<T_NodeLeft, T_NodeRight>::type*>(
+            addNext(static_cast<AstNode*>(nodep), static_cast<AstNode*>(newp)));
+    }
```

Comment:
I believe `common_type` does not give you the LUB type if the two types are siblings in a type hierarchy (see https://godbolt.org/z/oK18TP8cG), so this would be better as
```
    template <typename T_NodeResult, typename T_NodeNext>
    static T_NodeResult* addNext(T_NodeResult* nodep, T_NodeNext* nextp) {
        static_assert(T_NodeNext is subtype of T_NodeResult);
        // also static_assert(T_NodeResult is subtype of AstNode);
        return static_cast<T_NodeResult*>(...);
    }
```
which is slightly less flexible but covers the common use case. Same for `addNextNull`

Reply:
Right, fixed, though I had to use `enable_if` so that we default to the non-template version if `!(T_NodeNext is subtype of T_NodeResult)`.